From 66cc13d7427c0179ded175793b9a0f54229c3b91 Mon Sep 17 00:00:00 2001
From: Allen Ye <allen.ye@mediatek.com>
Date: Mon, 17 Feb 2025 20:56:51 +0800
Subject: [PATCH 75/95] mtk: mt76: mt7996: add remove key for bigtk

Fix sta beacon loss by removing the bigtk of previous bss installed in fw.

Connac3 fw can't remove group seperately, there is a way that driver can
use add action with bigtk key index and didn't specify cipher to tell fw to
disable sw mode bigtk.

Signed-off-by: Allen Ye <allen.ye@mediatek.com>
---
 mt7996/main.c | 20 +++++++++++++-------
 mt7996/mcu.c  | 16 ++++++++--------
 2 files changed, 21 insertions(+), 15 deletions(-)

diff --git a/mt7996/main.c b/mt7996/main.c
index cdc56061..eca854ca 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -717,6 +717,9 @@ static int mt7996_set_key(struct ieee80211_hw *hw, enum set_key_cmd cmd,
 	unsigned int link_id;
 	u8 pn[6] = {};
 
+	if (cmd != SET_KEY && cmd != DISABLE_KEY)
+		return -EINVAL;
+
 	if (key->link_id >= 0) {
 		add = BIT(key->link_id);
 	} else {
@@ -741,11 +744,12 @@ static int mt7996_set_key(struct ieee80211_hw *hw, enum set_key_cmd cmd,
 		conf = link_conf_dereference_protected(vif, link_id);
 		mconf = mt7996_vif_link(dev, vif, link_id);
 		msta_link = mt76_dereference(msta->link[link_id], &dev->mt76);
-		wcid_keyidx = &msta_link->wcid.hw_key_idx;
 
-		if (!conf || !mconf || !msta_link)
+		if ((cmd == SET_KEY && !conf) || !mconf || !msta_link)
 			continue;
 
+		wcid_keyidx = &msta_link->wcid.hw_key_idx;
+
 		/* fall back to sw encryption for unsupported ciphers */
 		switch (key->cipher) {
 		case WLAN_CIPHER_SUITE_TKIP:
@@ -775,7 +779,7 @@ static int mt7996_set_key(struct ieee80211_hw *hw, enum set_key_cmd cmd,
 			return -EOPNOTSUPP;
 		}
 
-		/* Todo: remove me after fix set dtim period to fw */
+		/* Necessary for fw cipher check */
 		if (cmd == SET_KEY && !sta && !mconf->mt76.cipher) {
 			mconf->mt76.cipher = mt76_connac_mcu_get_cipher(key->cipher);
 			mt7996_mcu_add_bss_info(mconf->phy, vif, conf, &mconf->mt76, msta_link, true);
@@ -783,12 +787,14 @@ static int mt7996_set_key(struct ieee80211_hw *hw, enum set_key_cmd cmd,
 
 		if (cmd == SET_KEY) {
 			*wcid_keyidx = idx;
-		} else {
-			if (idx == *wcid_keyidx)
-				*wcid_keyidx = -1;
-			goto out;
+		} else if (idx == *wcid_keyidx) {
+			*wcid_keyidx = -1;
 		}
 
+		/* To remove BIGTK independently, FW needs an extra inband command */
+		if (cmd == DISABLE_KEY && !(idx == 6 || idx == 7))
+			goto out;
+
 		mt76_wcid_key_setup(&dev->mt76, &msta_link->wcid, key);
 
 		err = mt7996_mcu_add_key(&dev->mt76, mconf, key,
diff --git a/mt7996/mcu.c b/mt7996/mcu.c
index f1c3466a..9ffef8cd 100644
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -3547,30 +3547,30 @@ mt7996_mcu_sta_key_tlv(struct mt76_dev *dev,
 		       u8 *pn)
 {
 	struct sta_rec_sec_uni *sec;
+	struct sec_key_uni *sec_key;
 	struct tlv *tlv;
 
 	tlv = mt76_connac_mcu_add_tlv(skb, STA_REC_KEY_V2, sizeof(*sec));
 	sec = (struct sta_rec_sec_uni *)tlv;
 	sec->add = cmd;
+	sec->n_cipher = 1;
+	sec_key = &sec->key[0];
+	sec_key->wlan_idx = cpu_to_le16(wcid->idx);
+	sec_key->key_id = key->keyidx;
 
 	if (cmd == SET_KEY) {
-		struct sec_key_uni *sec_key;
 		u8 cipher;
 
 		cipher = mt76_connac_mcu_get_cipher(key->cipher);
 		if (cipher == MCU_CIPHER_NONE)
 			return -EOPNOTSUPP;
 
-		sec_key = &sec->key[0];
-		sec_key->wlan_idx = cpu_to_le16(wcid->idx);
 		sec_key->mgmt_prot = 0;
 		sec_key->cipher_id = cipher;
 		sec_key->cipher_len = sizeof(*sec_key);
-		sec_key->key_id = key->keyidx;
 		sec_key->key_len = key->keylen;
 		sec_key->need_resp = 0;
 		memcpy(sec_key->key, key->key, key->keylen);
-		/* Todo: move the BCN protection logic into mt76_connac_mcu_get_cipher() */
 		if (sec_key->key_id == 6 || sec_key->key_id == 7) {
 			switch (key->cipher) {
 			case WLAN_CIPHER_SUITE_AES_CMAC:
@@ -3596,10 +3596,10 @@ mt7996_mcu_sta_key_tlv(struct mt76_dev *dev,
 			memcpy(sec_key->key + 16, key->key + 24, 8);
 			memcpy(sec_key->key + 24, key->key + 16, 8);
 		}
-
-		sec->n_cipher = 1;
 	} else {
-		sec->n_cipher = 0;
+		/* connac3 fw use set key action to apply removing bigtk and other
+		 * group keys should just use set key to overwrite the old ones. */
+		sec->add = SET_KEY;
 	}
 
 	return 0;
-- 
2.45.2

